### DISCUSSION TITLE
Where does screeni-py Fetches its data from?

### DISCUSSION THREAD

**Discussion by sachinkar** [pranjal-joshi/Screeni-py#113] [Q&A] [answered]
Hello. I just stumbled upon this interesting tool. Ran a few test queries and saw a line that says "Getting Stock Codes From NSE..." which got me wondering how does screeni-py get its data? Do you host a separate private server/DB where it sends all its queries and gets the data from or do you use NSE's API or something? I haven't gone through the source code yet so just figured I'd leave a question here in Q&A section.

**Comment by pranjal-joshi** [comment, marked as answer]
Hi @sachinkar,
The `Fetcher` class uses the CSV file from the old NSE website to get the group of stocks by their market cap. This only provides the Name of the stocks and not the OHLC data.
Please view the following source code to get more idea
https://github.com/pranjal-joshi/Screeni-py/blob/6f9bb0f0aa93ce90d9223bfcdd73d73cde1ae9da/src/classes/Fetcher.py#L37-L68

After getting the stock names, screenipy uses Yahoo finance (`yfinance`) API to get OHLC data as per the screening configuration during the market hours.
https://github.com/pranjal-joshi/Screeni-py/blob/6f9bb0f0aa93ce90d9223bfcdd73d73cde1ae9da/src/classes/Fetcher.py#L111-L118

In After-Market hours, the GitHub Actions automatically fetch & cache the closing data from the same API and the screenipy will be using this cache instead of re-fetching through `yfinance` because the market is not ticking any live data and cache is several times faster than pulling the data from the API.

**Comment by sachinkar** [comment]
Hello @pranjal-joshi, Thanks for your reply. I now have a better understanding of the tool.

As you mentioned, the Fetcher class uses the CSV file from the old NSE website to get the group of stocks by their market cap, I was wondering why not use the new website to do it, as we all know the old site's life if now numbered and soon to be deprecated and removed. I've complied the link source of the new site for you to use, here it is:

https://archives.nseindia.com/content/indices/ind_nifty50list.csv
https://archives.nseindia.com/content/indices/ind_niftynext50list.csv
https://archives.nseindia.com/content/indices/ind_nifty100list.csv
https://archives.nseindia.com/content/indices/ind_nifty200list.csv
https://archives.nseindia.com/content/indices/ind_nifty200list.csv
https://archives.nseindia.com/content/indices/ind_nifty500list.csv
https://archives.nseindia.com/content/indices/ind_niftysmallcap50list.csv
https://archives.nseindia.com/content/indices/ind_niftysmallcap100list.csv
https://archives.nseindia.com/content/indices/ind_niftysmallcap250list.csv
https://archives.nseindia.com/content/indices/ind_niftymidcap50list.csv
https://archives.nseindia.com/content/indices/ind_niftymidcap100list.csv
https://archives.nseindia.com/content/indices/ind_niftymidcap150list.csv

Furthermore, I also see that currently, screenipy is NSE only. How about adding BSE screening as well? We can call all the relevant data from BSE site as well for all the indices, as we can see on BSE's Bhav Copy page here: https://www.bseindia.com/markets/MarketInfo/BhavCopy.aspx
Direct link would be something like this: https://www.bseindia.com/download/BhavCopy/Equity/EQ_ISINCODE_100522.zip / 
https://www.bseindia.com/download/BhavCopy/Equity/EQ100522_CSV.ZIP and soo on.

As per the Index data, we can fetch it from the Historical Data page here: https://www.bseindia.com/Indices/IndexArchiveData.html

Direct link would be something like this: https://api.bseindia.com/BseIndiaAPI/api/ProduceCSVForDate/w?strIndex=SENSEX&dtFromDate=10/05/2022&dtToDate=10/05/2022

I haven't done proper research on these yet, but I am sure we can figure out faster and better ways if we put in our time.

**Reply by pranjal-joshi** [reply]
Great!
We can update the CSV URLs in the new version. Please go through the [Contribution](https://github.com/pranjal-joshi/Screeni-py/blob/main/CONTRIBUTING.md) guidelines given on the repo and feel free to make a contribution if possible.
It was a personal preference to consider NSE only as usually there is good volume and liquidity is observed on NSE as compared to BSE. Also, the Free account of [TradingView](https://in.tradingview.com/) supports live charts from NSE but 15 minute delayed charts from BSE.
There are currently no facilities for Index & F&O trading in the current version of Screeni-py (v1.35) but looking forward for a feature development in collaboration if possible.

**Reply by sachinkar** [reply]
Yes, I'd surely contribute. As I am new to python I am eager to learn. And since I am an active trader I think I can come in with a lot of new feature suggestions in the coming days as I use and test the tool more. Will keep in touch.

**Reply by pranjal-joshi** [reply]
CSV Links have been updated in b98f629
